torchdata/nodes/io/file_list.py

```diff
@@ -0,0 +1,138 @@
+    - Supports glob patterns for file matching
+    - Maintains state for checkpointing and resumption
+
+    Input: None (configured with uri_base and patterns at initialization)
+    Output: Dictionary with file URI and metadata
+        {
+            "data": "s3://bucket/key.txt",  # File URI as string
+            "metadata": {
+                "fs_protocol": "s3",        # Filesystem protocol
+                "source": "s3"             # Source type
+            }
+        }
+
+    Examples:
+        >>> # List local files
+        >>> node = FileLister("path/to/local/dir", ["*.txt", "*.json"])
+        >>> for item in node:
+        ...     print(item["data"])  # prints "path/to/local/dir/file1.txt", etc.
+        >>>
+        >>> # List S3 files with anonymous access
+        >>> node = FileLister("s3://my-bucket/prefix", ["*.csv"],
+        ...                   fs_options={"anon": True})
+        >>> for item in node:
+        ...     print(item["data"])  # prints "s3://my-bucket/prefix/file1.csv", etc.
+        >>>
+        >>> # List GCS files
+        >>> node = FileLister("gs://my-bucket/prefix", ["*.parquet"])
+        >>> for item in node:
+        ...     print(item["data"])  # prints "gs://my-bucket/prefix/file1.parquet", etc.
+    """
+
+    SOURCE_KEY = "source"
+    CURRENT_IDX_KEY = "current_idx"
+
+    def __init__(self, uri_base: str, patterns: List[str], fs_options: Optional[Dict[str, Any]] = None):
+        """Initialize the FileLister.
+
+        Args:
+            uri_base: Base URI for file listing (any fsspec-supported URI)
+            patterns: List of glob patterns to match files against
+            fs_options: Options to pass to fsspec filesystem
+        """
+        super().__init__()
+
+        self.uri_base = uri_base
+        self.patterns = patterns if patterns else ["*"]
+        self.fs_options = dict(fs_options or {})
+
+        # Parse protocol from URI using fsspec
+        self.protocol, _ = fsspec.core.split_protocol(uri_base)
+        if self.protocol is None:
+            self.protocol = "file"  # Default to local filesystem
+
+        # Create filesystem instance
+        self.fs = fsspec.filesystem(self.protocol, **self.fs_options)
+
+        # Initialize state
+        self._file_paths = None
+        self._current_idx = 0
+
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
+        """Reset node state and populate file paths."""
+        super().reset(initial_state)
+
+        # Always initialize files list
+        self._file_paths = set()
+
+        try:
+            # Use fsspec's glob for file discovery
+            for pattern in self.patterns:
+                pattern_path = os.path.join(self.uri_base, pattern)
+
+                try:
+                    matches = self.fs.glob(pattern_path)
+
+                    # Filter out directories from matches if pattern is **/* or *
+                    if pattern == "**/*" or pattern == "*":
```

Comment:
what if the pattern is "test_files/*"
will it also return directories? Do we need to filter them?

Reply:
it will return files only. actually the next line is the filter for dirs.

`matches = [match for match in matches if self.fs.isfile(match)]`